pkg/debounce/debounce.go

```diff
@@ -13,75 +13,90 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+	Input <-chan T
 
-// Debouncer defines the debouncer interface
-type Debouncer interface {
-	Stop()
-	Start()
-}
+	// Timeout defines the time that must pass without seeing any new items
+	// until the most recent item is forwarded to the callback.
+	Timeout time.Duration
 
-type debouncer struct {
-	ctx      context.Context
-	cancel   context.CancelFunc
-	interval time.Duration
-	callback Callback
-	input    chan fsnotify.Event
-}
+	// Filter controls which elements reset the time window and which elements
+	// are are simply silently dropped. It may be nil, in which case no items
+	// will be dropped.
+	Filter func(item T) bool
 
-// New creates new Debouncer with given args
-func New(ctx context.Context, interval time.Duration, input chan fsnotify.Event, callback Callback) Debouncer {
-	dctx, cancel := context.WithCancel(ctx)
-	db := &debouncer{
-		cancel:   cancel,
-		interval: interval,
-		input:    input,
-		ctx:      dctx,
-		callback: callback,
-	}
-
-	return db
+	// Callback is the func that receives debounced items.
+	Callback func(item T)
 }
 
-// Start starts the debouncer
-func (d *debouncer) Start() {
-	ticker := time.NewTimer(d.interval)
-	var item *fsnotify.Event
-	for {
+// Run polls the input for items, throttles them and forwards them to the
+// callback. Run returns either without an error when the input is closed or
+// with the context's error if the context is done, whichever happens first.
+func (d *Debouncer[T]) Run(ctx context.Context) error {
+	// Create a timer that is initially expired
+	timer := time.NewTimer(0)
+	defer timer.Stop()
+	<-timer.C
+
+	for pendingItem := (*T)(nil); ; {
 		select {
-		case event := <-d.input:
-			if event.Op != fsnotify.Chmod {
-				item = &event
-				ticker.Reset(d.interval)
+		case <-ctx.Done():
+			return ctx.Err() // ctx is done, good bye ...
+
+		case item, ok := <-d.Input:
+			if !ok {
+				return nil // input channel closed, good bye ...
 			}
-		case <-ticker.C:
-			if item != nil {
-				d.callback(*item)
+
+			if d.Filter != nil && !d.Filter(item) {
+				continue // the current item has been filtered out ...
 			}
-		case <-d.ctx.Done():
-			return
+
+			// The current item needs to reset the timer.
+			// There are three possible states:
+			// https://stackoverflow.com/a/58631999
+			if pendingItem == nil {
+				// (1) The pendingItem has been consumed.
+				// The timer must have expired and its tick must have been consumed.
+			} else if timer.Stop() {
+				// (2) The timer was stopped before it expired.
+				// Its channel cannot hold a tick.
+			} else {
+				// (3) The timer has already expired, but the pendingItem has not been consumed yet.
+				// The pending tick needs to be consumed.
+				<-timer.C
+			}
+
+			timer.Reset(d.Timeout)
+
+			// Save the current item to be sent after the timeout.
+			pendingItem = &item
+
+			// Now it is safe to restart the timer.
+			timer.Reset(d.Timeout)
+
+		case <-timer.C:
+			d.Callback(*pendingItem)
```

Comment:
Is it possible to get the timer before the input? I'm looking at the dereferencing of `pendingItem` here.

Reply:
It shouldn't be possible; I consider reaching this code here when `pendingItem` is `nil` a logic error.

When the select loop enters, it will enter with `pendingItem == nil` and an expired timer and this branch won't be reached. When there's a new item stored into `pendingItem`, the previous branch ensures that this branch will fire exactly 0 or 1 times, depending on the time between the incoming items. The previous branch stores the current item into `pendingItem` after it resets the timer, but since the timer channel is only polled in the exact same loop, its tick is guaranteed to be consumed after `pendingItem` has been updated.